File: ocaml/lib/client_pool.ml
module ClientPool = struct
  open Lwt.Infix
  
  module LRU = Lru.F.Make(String)
  
  type 'a t = {
    client_cache: (string, 'a) LRU.t Lwt_mutex.t;
  }
  
  let client_cache_size = 20
  
  module type Connectable = sig
    type t
    val connect : string -> t Lwt.t
  end
  
  let create () = 
    { client_cache = Lwt_mutex.create (LRU.empty client_cache_size) }
  
  let connect (type a) (module C : Connectable with type t = a) pool address =
    C.connect address >>= fun client ->
    Lwt_mutex.with_lock pool.client_cache (fun cache ->
      let cache = LRU.add address client cache in
      Lwt.return (client, { pool with client_cache = Lwt_mutex.create cache })
    )
  
  let get_client (type a) (module C : Connectable with type t = a) pool address =
    Lwt_mutex.with_lock pool.client_cache (fun cache ->
      match LRU.find address cache with
      | Some client -> Lwt.return (client, pool)
      | None -> 
          C.connect address >>= fun client ->
          let cache = LRU.add address client cache in
          Lwt.return (client, { pool with client_cache = Lwt_mutex.create cache })
    )
end

File: haskell/lib/Client_pool.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE FlexibleContexts #-}

module ClientPool 
  ( ClientPool
  , Connectable(..)
  , create
  , connect
  , getClient
  ) where

import Control.Concurrent.STM
import Control.Monad.IO.Class
import Data.Cache.LRU as LRU
import Data.Text (Text)
import qualified Data.Text as T

data ClientPool a = ClientPool
  { clientCache :: TVar (LRU Text a)
  }

class Connectable a where
  connect :: Text -> IO a

clientCacheSize :: Int
clientCacheSize = 20

create :: IO (ClientPool a)
create = ClientPool <$> newTVarIO (LRU.newLRU (Just clientCacheSize))

connectClient :: (MonadIO m, Connectable a) => ClientPool a -> Text -> m a
connectClient pool address = liftIO $ do
  client <- connect address
  atomically $ do
    cache <- readTVar (clientCache pool)
    let (_, cache') = LRU.insert address client cache
    writeTVar (clientCache pool) cache'
  return client

getClient :: (MonadIO m, Connectable a) => ClientPool a -> Text -> m a
getClient pool address = liftIO $ atomically $ do
  cache <- readTVar (clientCache pool)
  case LRU.lookup address cache of
    Just client -> return client
    Nothing -> do
      client <- liftIO $ connect address
      let (_, cache') = LRU.insert address client cache
      writeTVar (clientCache pool) cache'
      return client